src/server/utils/file.js

```diff
@@ -0,0 +1,150 @@
+import { promises as fs } from "fs";
+import path from "path";
+import slugify from "slugify";
+import { loadFront } from "yaml-front-matter";
+
+export async function getFiles(dir, recursive = true) {
+  const dirents = await fs.readdir(dir, { withFileTypes: true });
+  const files = await Promise.all(
+    dirents.map((dirent) => {
+      const res = path.resolve(dir, dirent.name);
+
+      if (!recursive && dirent.isDirectory()) return [];
+
+      return dirent.isDirectory() ? getFiles(res) : res;
+    })
+  );
+
+  return files.flat();
+}
+
+export async function processMdFile(dir, filePath) {
+  const pathKey = path
+    .relative(dir, path.dirname(filePath))
+    .split(path.sep)
+    .join("/");
+
+  const rawContent = await fs.readFile(filePath, { encoding: "utf8" });
+  const { __content: content, ...metadata } = loadFront(rawContent);
+  const fileName = path.parse(filePath).name;
+  const route = [pathKey, fileName].filter(Boolean).join("/");
+
+  return {
+    filePath,
+    fileName,
+    pathKey,
+    content,
+    metadata,
+    route,
+  };
+}
+
+export async function getMarkdownPages(dir) {
+  const files = await getFiles(dir);
+  const pages = await Promise.all(files.map((f) => processMdFile(dir, f)));
+
+  // Sort pages by pathKey and metadata order
+  pages.sort((a, b) => {
+    if (a.pathKey === b.pathKey)
+      return Math.sign(a.metadata?.order - b.metadata?.order);
+
+    if (a.pathKey.startsWith(b.pathKey)) return -1;
+    if (b.pathKey.startsWith(a.pathKey)) return 1;
+
+    return 0;
+  });
+
+  return pages;
+}
+
+const LINE_BREAK_REGEX = /[\r\n]+/;
+const HEADING_REGEX = /^(?<prefix>#{1,6})\s(?<name>.*)$/i;
+
+const findClosestParent = (depth, startingParent) => {
+  let parent = startingParent;
+
+  while (depth <= parent.depth) {
+    parent = parent.parent;
+  }
+
+  return parent;
+};
+
+const removeLinks = (node) => {
+  delete node.parent;
+
+  node.children.forEach(removeLinks);
+};
+
+const defaultSlugify = (str) => slugify(str, { lower: true });
+const defaultCreateId = (slug, depth) => slug;
+
+export const createFileTOC = async (filePath, options = {}) => {
+  const {
+    name = "",
+    linkRoot = "/",
+    createSlug = defaultSlugify,
+    maxDepth = Infinity,
+    id = defaultCreateId,
+  } = options;
+  const md = await fs.readFile(filePath, { encoding: "utf8" });
+  const lines = md.split(LINE_BREAK_REGEX).map((line) => line.trim());
+  const slug = createSlug(name);
+
+  const rootParent = {
+    id: id(slug, 0),
+    name,
```

Comment:
If I'm understanding correctly, the rootParent `name` is set by options. If it's not given in options it should default to the first heading string rather than an empty string

Reply:
Or the filename / directory name

Reply:
Sure, directory name is easy enough. The first heading bit is do-able but I'd like to spend some more time thinking about that first.